Implemented Gallery Drag and Drop Reordering

caxco93 opened the pull request:
## Description
    Implemented Gallery Drag and Drop Reordering
    * Created new Sortable HOC for wrapping the Gallery Images
    * Added react-sortable-hoc dependency

Addresses Issue #743

## Screenshots
![ezgif-1-88a02762a5](https://user-images.githubusercontent.com/3190666/43881861-3c9e9c8a-9b73-11e8-8d9f-a0158a14b676.gif)

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [X] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

- [ ] My code is tested.
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->

## KNOWN ISSUES:
    * Selected image styles are not displayed properly while dragged (this is because
      react-sortable-hoc creates a copy of the dragged element right before the </body>
      tag, but for these images, the style depend on the <ul> parent)

caxco93 committed:
Implemented Gallery Drag and Drop Reordering

* Created new Sortable HOC for wrapping the Gallery Images

* Added react-sortable-hoc dependency

* Right now you have to hold down to reorder

caxco93 committed:
Properly extracted prop function

caxco93 commented:
Yeah having it as a generic independent component would be ideal. Right now it's being specific to this block behavior though. I think after its completed it can be extracted.

caxco93 commented:
Hey everyone I was wondering whether the reordering behaviour should work with a hold press delay to start dragging, or a handle from where you can drag instantaneously. What do you think?

SuperGeniusZeb commented:
Yeah, I would prefer instant drag as well, unless there is some technical issue that prevents it.

caxco93 commented:
gotcha! i'll try to get it to work that way then!

caxco93 committed:
Fixed the missing "Upload an image"

Altered the Sortable component so it now handles custom first and last
nodes.

caxco93 committed:
Changed drag behaviour so it 'just works'

caxco93 committed:
Added missing class to the dragged element

Addresses the floating square when dragging

caxco93 commented:
Right now im dealing with the Image Caption Focus resetting after 1 key stroke. I have just now noticed it has been caused by the HOC around the GalleryImage components.
React is losing track of the latter and thus unable to preserve the state.

caxco93 committed:
Added shouldComponentUpdate to prevent unncessary renders

caxco93 committed:
Modified imports to use proper WordPress packages

noisysocks commented:
I'm getting an exception when I:

1. Insert a new gallery block
2. Click _Media Library_
3. Select some images, then proceed to _Insert gallery_

```
sortable.js:39 Uncaught TypeError: Cannot add property tabindex, object is not extensible
    at sortable.js:39
    at Array.map (<anonymous>)
    at sortable.js:38
    at mountIndeterminateComponent (react-dom.24169eaf.js:13958)
    at beginWork (react-dom.24169eaf.js:14398)
    at performUnitOfWork (react-dom.24169eaf.js:16441)
    at workLoop (react-dom.24169eaf.js:16480)
    at HTMLUnknownElement.callCallback (react-dom.24169eaf.js:140)
    at Object.invokeGuardedCallbackDev (react-dom.24169eaf.js:178)
    at invokeGuardedCallback (react-dom.24169eaf.js:227)
```

noisysocks commented on package.json:
```diff
@@ -55,6 +55,7 @@
 		"re-resizable": "4.7.1",
 		"react": "16.4.1",
 		"react-dom": "16.4.1",
+		"react-sortable-hoc": "0.8.3",
```
How many bytes does this add to our bundle size?

caxco93 replied:
Around 40 KB

caxco93 committed:
Removed unnecessary prop mutation that raised a warning

caxco93 committed:
Extracted Sortable HOC

Got rid of closure higher order component in favor of readability

caxco93 commented:
Thanks for spotting that. It should be addressed now.

caxco93 commented:
This `react-sortable-hoc` is raising some React warnings though. And I am also starting to notice some performance issues. Trying to decide if we should keep using it or maybe roll our own. I originally intended to use the HTML5 Drag and Drop API but had problems with the GalleryImages being nested inside another draggable.

ehg commented:
> Trying to decide if we should keep using it or maybe roll our own.
What are the trade-offs between using it or rolling our own?

ehg commented:
There are some code formatting/style issues that should be fixed. Are you running a linter?

ehg commented on core-blocks/gallery/sortable.js:
```diff
@@ -0,0 +1,119 @@
+const SortableItem = SortableElement( ( { children } ) => children );
+
+class Sortable extends Component {
+	//constructor
```
Why comment here?

ehg replied:
Also, we could get rid of the constructor all together using by using class properties.

caxco93 replied:
You are right let me fix that

caxco93 replied:
we don't have support for class properties enabled though, and other blocks/components seem to be relying on binding inside the constructor as well. Should we enable it?

noisysocks replied:
We're reluctant to use new language features until they've reached stage 4. [Class fields are still at stage 3](https://github.com/tc39/proposal-class-fields).

caxco93 commented:
Yes I am using eslint on vscode. You mean the function comments right? Thanks for noticing that. I'll check and solve it

caxco93 committed:
Fixed minor code formatting/style issues

caxco93 commented:
@ehg Rolling our own one or a fork would let us modify the component that is being rendered while dragging and also address the performance issues directly; however, it would take more time. 
Those are the two main missing points by now.

caxco93 committed:
Corrected misbehaving selection after drop reorder

caxco93 commented:
Hey, so I addressed the sizing issue when reordering. It now looks nice like this:
![dndflowresize2](https://user-images.githubusercontent.com/3190666/44704226-2cccc600-aa60-11e8-8737-eec207278d71.gif)

I was thinking I would need to set a class on the rearranging elements until I realized the dynamic size of the gallery items is not caused by a different class, but by their `flex-grow` property and their DOM index position inside the flex-wrappable parent.
The way I did it is by calculating the correspondent scale factor given the ratio of the different item dimensions 👍 

However, this behaviour has been achieved by modifying the dependency library `react-sortable-hoc` locally.

I created a PR there (https://github.com/clauderic/react-sortable-hoc/pull/422), but he doesn't really look that active right now. 👎

nosolosw commented:
Hey, just wanted to note that if it crossed your mind to get rid of `react-sortable-hoc` and implement this behavior within Gutenberg, you may take advantage of the `Draggable` component. It's being refactored at https://github.com/WordPress/gutenberg/pull/9311 to enabling its use in use cases other than drag blocks around (for example, dragging pictures!) - it's almost ready to be merged.

muhammedmagdi commented:
@nosolosw  Can you just provide us with full example for how to use the `Draggable` component in some use cases other than drag blocks.

nosolosw commented:
@muhammedmagdi the PR I've mentioned also updates the docs, see [`README`](https://github.com/WordPress/gutenberg/blob/53b10d3ffec11b4d6eacc57f98639eab86852ce4/packages/components/src/draggable/README.md) for an example. See also the [`BlockDraggrable`](https://github.com/WordPress/gutenberg/blob/53b10d3ffec11b4d6eacc57f98639eab86852ce4/packages/editor/src/components/block-list/block-draggable.js) component for a concrete implementation, it makes use of `Draggable`.

That PR is now approved and it's going to be merged into master next week, after the 3.7 Gutenberg release. From that point on, it'll be available for anyone to use.

muhammedmagdi commented:
@nosolosw 
Thank you.

caxco93 committed:
Fixed styling issue on currently dragged element

The Gallery Item Image did not have proper styling because
of the way the dragged element is created (outside its <ul>)
This adds the necessary css to address this.

gziolo commented:
@mapk and @youknowriad - what's the plan regarding this feature for Phase 2?

youknowriad commented:
@gziolo It just needs dev :)

nosolosw commented:
@gziolo @youknowriad I had this on my radar to work on after my the doc generator task. We may want to close this PR and reference the master issue https://github.com/WordPress/gutenberg/issues/743 everywhere (it contains other references). I'll just do that. If we end up using this approach, we can always reopen.

gziolo commented:
Awesome, thanks @nosolosw for making it clear. Indeed, we can always reopen as we get back to this issue. Let's keep the number of open PRs in a healthy state :)
